### PR TITLE
ci: optimizations and ccache support

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# Code Review Checklist

## Purpose

CI pipeline optimizations.

Key changes:
* Add `ccache` support for LLVM builds speedining up re-compilation up to 90%
* Report `lit` tests in `xunit` format as part of PR check as additional comment to have visibility of executed unit tests
* Combine code duplications in one `build` action and re-use across the workflows
* Simplify workflows while keeping the same functionality
* Use open source `matterlabs/llvm_runner:ubuntu22-llvm15-latest` image without private credentials

## Ticket Number

N/A

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: .github/workflows/compiler-integration-tests.yml [hunk 1]
```diff
@@ -176,27 +142,34 @@ jobs:
...
 
       - uses: 8398a7/action-slack@v3
         with:
           status: ${{ job.status }}
-          fields: repo,commit,author,action,eventName,ref,workflow,job,took,pullRequest # selectable (default: repo,message)
+          fields: repo,commit,author,action,eventName,ref,workflow,job,took,pullRequest
         env:
-          SLACK_WEBHOOK_URL: ${{ secrets.MATTERMOST_WEBHOOK }} # required
-        if: always() # Pick up events even if the job fails or is canceled.
+          SLACK_WEBHOOK_URL: ${{ secrets.MATTERMOST_WEBHOOK }}
```
Comment (hedgar2017) [inline .github/workflows/compiler-integration-tests.yml]: Is this mattermost supposed to be replaced?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]

File: .github/workflows/compiler-integration-tests.yml (hunk 1 above)
Comment (antonbaliasnikov) [reply .github/workflows/compiler-integration-tests.yml]: To be discussed, probably, we need to open a new ITS ticket for testing webhook, because the one we've got is only for release notifications.

Comment (hedgar2017) [reply .github/workflows/compiler-integration-tests.yml]: @antonbaliasnikov understood.
Could you please open it? At least to clarify these namings first.

Comment (antonbaliasnikov) [reply .github/workflows/compiler-integration-tests.yml]: I believe we could just remove these as we agree in Slack?

This is job notifications defined for all PRs sent in one place in Slack, I don't think it's useful that way. What we can implement later is to run workflows for `main` and report only failures happened there.

Comment (hedgar2017) [reply .github/workflows/compiler-integration-tests.yml]: Let's do it this way:
https://linear.app/matterlabs/issue/CPR-1586/fix-ci-reports-that-are-sent-to-info-compiler-ci

**Review by hedgar2017 — APPROVED** [review]

File: .github/actions/build/action.yml [hunk 2]
```diff
@@ -25,6 +25,11 @@ runs:
       run: brew install cmake ninja
       shell: ${{ runner.os == 'Windows' && 'msys2 {0}' || 'bash' }}
 
+    - name: Prepare ccache installation
+      if: runner.os == 'Linux'
+      shell: ${{ runner.os == 'Windows' && 'msys2 {0}' || 'bash' }}
```
Comment (hedgar2017) [inline .github/actions/build/action.yml]: Just curious, why do we need this Windows thing here?

**Review by hedgar2017 — APPROVED** [review]

File: .github/actions/build/action.yml (hunk 2 above)
Comment (antonbaliasnikov) [reply .github/actions/build/action.yml]: Good catch, for this special case it should be just `bash` here, I'll fix

**Comment by antonbaliasnikov** [comment]
Compiler benchmarks failure is expected, because there is no bug fix for the reference branch. It's tested with the previous version, so, it's safe to merge if integration tests all pass.

**Comment by hedgar2017** [comment]
@antonbaliasnikov the `candidate` was canceled for some reason. Is it OK?

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov the `candidate` was canceled for some reason. Is it OK?

Yes, `reference` failed and `candidate` was cancelled as nothing to compare the results to.

**Comment by hedgar2017** [comment]
@antonbaliasnikov gotcha, could you squash?

**Comment by antonbaliasnikov** [comment]
> @antonbaliasnikov gotcha, could you squash?

Done. I believe we can merge. Are we always `Rebase and merge` here with manual squashes?

**Comment by hedgar2017** [comment]
> > @antonbaliasnikov gotcha, could you squash?
> 
> Done. I believe we can merge. Are we always `Rebase and merge` here with manual squashes?

Yes, it's to keep linear history in LLVM.
Unfortunately, GH doesn't provide `Squash and rebase and merge`, unless I'm missing something.

**Comment by antonbaliasnikov** [comment]
> > > @antonbaliasnikov gotcha, could you squash?
> > 
> > 
> > Done. I believe we can merge. Are we always `Rebase and merge` here with manual squashes?
> 
> Yes, it's to keep linear history in LLVM. Unfortunately, GH doesn't provide `Squash and rebase and merge`, unless I'm missing something.

@hedgar2017 , I believe I was able to achieve this by enabling `main` branch protection rule:
image

And then configuring PR merges like that:
image
image

**Comment by hedgar2017** [comment]
@antonbaliasnikov cool, let's try it out next time.
